### DISCUSSION TITLE
Jinja template rendering not happening when operators are nested

### DISCUSSION THREAD

**Discussion by mishrasangeeta87** [apache/airflow#38311] [Q&A] [answered]
Hello Everyone, 

I'm reaching out regarding a use case where I need to pass additional metadata to all my operators. To address this, I've implemented a decorator over the PythonOperator and nested other operators under it to extract the metadata. However, I'm currently encountering an issue. Within this Python operator, when I invoke another operator such as BashOperator or KubernetesPodOperator that contains template variables, the variables resolve properly when the nested operator isn't encapsulated within the Python operator. However, once I nest it within the Python Operator, the template resolution fails to occur.

Additionally, I've noticed that the nested operator's execute method isn't called automatically; I have to call it explicitly.

I would greatly appreciate any insights or pointers you may have regarding this matter.

Below is the code snippet:
```


def standard_wrapped_task(**context):
    standard_dbt_bas_task = bash_operator_factory.get_dummy_dbt_bash_task(name='standard-bash-task',
                                                                          dbt_cmd="echo dbt_cmd_run",
                                                                          selector='--models tag:standard', **context)
    standard_dbt_bas_task.execute(context)
    return standard_dbt_bas_task


standard_task = PythonOperator(
    task_id="standard-python-task",
    python_callable=standard_wrapped_task,
    provide_context=True,
    dag=dag
)
```

**Comment by potiuk** [comment, marked as answer]
There is no concept of "nested" operators in Airflow. Not sure where you took the idea that you can do it. This is not going to work - becuase operators can only be used and executed by the Airflow framework - if you are creating operator inside execute method of airflow , you are merely creating an object and calling it's execute method - it does not actually "run" the operator.

We noticed that somehow Airlfow users **think** that this is something that is possible, but it's not - you can read discussion in devlist here https://lists.apache.org/thread/nflt9h6dc5obzztmyqxlpxfs950rtqsq and if you have your own experience and comments and feel that this case should have some special support and that the ways we explained people can deal with similar issues is not enough, feel free to comment in the discussion.

We also currently review and are about to merge and relase in 2.9.0 a protection against such cases - where any attempts to do what you attempt to do will produce an error explaining that you should not do it: https://github.com/apache/airflow/pull/37937

**Comment by potiuk** [comment]
BTW. Is this possible to know WHERE you took the idea from about this way of using "nested operators" ? I am really curious to hear if there is a source of information that tells you can do it, or is it something you invented on your own ? Or maybe that was suggested by some AI /Chat GPT like question/answer maybe?

**Reply by mishrasangeeta87** [reply]
Thanks @potiuk  for the detailed explanation.  I came up with this idea on my own to reuse the logic written in python and extend it for other operators as well.

**Comment by kyleburke-meq** [comment]
@mishrasangeeta87 you can render them in your outer function function before passing it into your nested function like this:

```python
def standard_wrapped_task(**context):
    rendered_jinja_template = context["task"].render_template({{ whatever you need to render here }}, context)
    standard_dbt_bas_task = bash_operator_factory.get_dummy_dbt_bash_task(name='standard-bash-task',
                                                                          dbt_cmd="echo dbt_cmd_run",
                                                                          selector='--models tag:standard', **context)
    standard_dbt_bas_task.execute(context)
    return standard_dbt_bas_task
```

**Comment by fritz-astronomer** [comment]
Airflow Hooks **should** contain the functionality that the operator is using, which you can use in your functions, or you can extend the operator
```
class MyCustomerOperator(MyOperator):
  ...
  ```

**Reply by fritz-astronomer** [reply]
then rendering will work as expected. As per @potiuk 's comment - Operators should be executed by Airflow, not custom code. There's overhead that is required.
